# Indexed-TGIRT-Seq

This folder contain scripts for tag-based error correction of TGIRT-seq developed by [Douglas Wu](mailto:wckdouglas@gmail.com).

Scripts:

* readClusterPairs.py:   Writing paired fastq files with concensus base from a cluster of reads. 
* pileupBam.py:         pileup bam file and  count the bases at each position column (depends on pysam) 



### readClusterPairs.py
```
usage: readClusterPairs.py [-h] -o OUTPUTPREFIX -1 FASTQ1 -2 FASTQ2
                           [-m CUTOFF] [-x IDXBASE] [-q BARCODECUTOFF] [-n]
                           [-t THREADS]

Clustering fastq reads to fasta reads with the first $IDXBASE bases as cDNA-
synthesis barcode. Concensus bases are called only when the fraction of reads
that contain the concensus base exceed some threshold. Quality scores are
generated by the average score for the bases that matched concensus base.

optional arguments:
  -h, --help            show this help message and exit
  -o OUTPUTPREFIX, --outputprefix OUTPUTPREFIX
                        Paired end Fastq files with R1_001.fastq.gz as suffix
                        for read1, and R2_001.fastq.gz as suffix for read2
  -1 FASTQ1, --fastq1 FASTQ1
                        Paired end Fastq file 1 with four line/record
  -2 FASTQ2, --fastq2 FASTQ2
                        Paired end Fastq file 2 with four line/record
  -m CUTOFF, --cutoff CUTOFF
                        minimum read count for each read cluster (default: 4)
  -x IDXBASE, --idxBase IDXBASE
                        how many base in 5' end as index? (default: 13)
  -q BARCODECUTOFF, --barcodeCutOff BARCODECUTOFF
                        Average base calling quality for barcode sequence
                        (default=30)
  -n, --retainN         Use N-containing sequence for concensus base vote and
                        output sequences containing N (default: False)
  -t THREADS, --threads THREADS
                        Threads to use (default: 1)
```

### pileupBam.py
```
usage: pileupBamToBase.py [-h] -i BAMFILE [-q QUALTHRESH] -r REFFASTA
                          [-d DEPTH] [-p THREADS] [-s SKIPBASES]

Pile up bam file and get mapped bases (excluded clipped bases and low quality
base).

optional arguments:
  -h, --help            show this help message and exit
  -i BAMFILE, --bamfile BAMFILE
                        position sorted bam file
  -q QUALTHRESH, --qualThresh QUALTHRESH
                        Base calling quality threshold (default: 33)
  -r REFFASTA, --refFasta REFFASTA
                        reference fasta
  -d DEPTH, --depth DEPTH
                        Maximum depth (default: 8000)
  -p THREADS, --threads THREADS
                        Threads to used (default: 1)
  -s SKIPBASES, --skipBases SKIPBASES
                        Bases at the end positions that will not be evaluated
                        (default: 3)
```
