# Indexed-TGIRT-Seq

This folder contain scripts for tag-based error correction of TGIRT-seq developed by [Douglas Wu](mailto:wckdouglas@gmail.com).

Scripts:

### read_cluster_pairs.py ###
This is the main script for calling the error-reducing program for paired-end sequences. This should be done before trimming to make it to work.

```
usage: read_cluster_pairs.py [-h] -o OUTPUTPREFIX -1 FASTQ1 -2 FASTQ2
                             [-m CUTOFF] [-x IDXBASE] [-q BARCODECUTOFF]
                             [-c CONSTANT_REGION] [-t THREADS] [-a MISMATCH]
                             -r {read1,read2}

Clustering fastq reads to fasta reads with the first $idx_base bases as cDNA-
synthesis barcode. Concensus bases are called only when the fraction of reads
that contain the concensus base exceed some threshold. Quality scores are
generated by the average score for the bases that matched concensus base.

optional arguments:
  -h, --help            show this help message and exit
  -o OUTPUTPREFIX, --outputprefix OUTPUTPREFIX
                        Paired end Fastq files with R1_001.fastq.gz as suffix
                        for read1, and R2_001.fastq.gz as suffix for read2
  -1 FASTQ1, --fastq1 FASTQ1
                        Paired end Fastq file 1 with four line/record
  -2 FASTQ2, --fastq2 FASTQ2
                        Paired end Fastq file 2 with four line/record
  -m CUTOFF, --cutoff CUTOFF
                        minimum read count for each read cluster (default: 4)
  -x IDXBASE, --idxBase IDXBASE
                        how many base in 5' end as index? (default: 13)
  -q BARCODECUTOFF, --barcodeCutOff BARCODECUTOFF
                        Average base calling quality for barcode sequence
                        (default=30)
  -c CONSTANT_REGION, --constant_region CONSTANT_REGION
                        Constant sequence after tags (default: CATCG ,e.g.
                        Douglas's index-R1R)
  -t THREADS, --threads THREADS
                        Threads to use (deflaut: 1)
  -a MISMATCH, --mismatch MISMATCH
                        Allow how many mismatch in constant region (deflaut:
                        1)
  -r {read1,read2}, --read {read1,read2}
                        barcode on first N bases of which read from pair end
```


### supplementary notes ###

A notes that support the software can be accessed [here](https://htmlpreview.github.io/?https://raw.githubusercontent.com/wckdouglas/indexed-tgirt-seq/master/notes/tagBased-Error.html?token=AGGQnAIgDPPQI30g-7AYkwhIsJMcEs4lks5XpNEKwA%3D%3D)

### Requirement ###
python =2.7
cython =>0.24
scipy =>0.18
numpy => 1.11
cjson =>1.0.5
biopython =>1.68
matplotlib =>1.5.1
seaborn =>0.7.1

